text/0000-builder-naming.md

```diff
@@ -0,0 +1,136 @@
+   the future, and the module itself has a descriptive name, calling the builder “Builder” is
+   recommended. For example:
+
+   * `thread::Builder` builds and spawns threads;
+   * `process::Builder` builds and spawns processes;
+
+   but *not* `fs::Builder`, which opens files;
+* If the builder produces one type of objects, use the name of buildee suffixed with “Builder”.
+    This scheme has an advantage of being searchable and self-documenting. For example:
+    `FileBuilder` is a builder and produces objects of type `File`;
+* If there is no clear buildee (most likely because it contains multiple finalizers), a noun
+    describing all the buildable objects suffixed with “Builder” may be used. For example:
+    `ThreadBuilder`, even though it produces `JoinHandle`s and `JoinGuard`s rather than `Thread`s;
+* Finally, if the majority (community) decides there exists a superior name for the builder, but it
+    does not match any of the guidelines above, generously point out the existence of the builder
+    in the documentation of buildee and module, if appropriate.
+
+### Methods
+
+Option setters – methods which represent options that can be set on the builder – should receive
+the builder object as `&mut self` rather than `self` unless there is a good reason to do otherwise.
+
+Finalizers – methods which produce the built object – should receive self via reference, unless
+there is a good reason to do otherwise.
+
+In order for method listing in documentation to be predictable, methods in the builder’s `impl`
+block should be declared following this order:
+
+* `new` – a method to create the builder object comes first in the `impl` block;
+* followed by one or more option setters;
+* followed by one or more finalizers.
+
+This ordering stems from conventional use of the builder pattern: create a builder, set the
+options and call a finalizer.
+
+#### Method naming
+
+Option setters should be nouns or adjectives. Avoid `is_*` prefix and nouns derived from
+adjectives. For example:
+
+* `name`;
+* `stack_size`;
+* `read` or `readable` (rather than `is_readable` or `readability`);
+* `write` or `writable` (rather than `is_writable` or `writability`).
+
+Finalizers should be verbs. For example:
+
+* `spawn`;
+* `create`;
+* `open`.
+
+## Proposed changes to the standard library
+
+### `thread::Builder`
+
+`thread::Builder` is adjusted to have the following interface:
+
+```rust
+pub struct Builder { … }
+impl Builder {
+    fn new() → Builder;
+    fn name(&mut self, name: String) → &mut Builder;
+    fn stack_size(&mut self, size: usize) → &mut Builder;
+    fn spawn<F>(&mut self, f: F) → Result<JoinHandle> where F: FnOnce(), F: Send + 'static;
+    fn scoped<'a, T, F>(&mut self, f: F) → Result<JoinGuard<'a, T>>
+        where T: Send + 'a, F: FnOnce() → T, F: Send + 'a
+}
+```
+
+and `impl Clone for Builder` is provided.
+
+In summary, all methods are changed to take and return `Builder` by mutable reference rather than
```

Comment:
Currently this is done because the `name` is an owned value (e.g. `scoped`/`spawn` consumes it), but I do think that this is largely a holdover from the past where `Box<Writer>` or `Box<FnOnce>` was contained within and the by-value-ness was needed.